src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs

```diff
@@ -6527,7 +6527,7 @@ void N()
         }
     }
 }",
-                MainDescription($"({FeaturesResources.local_variable}) string s"),
+                MainDescription($"({FeaturesResources.local_variable}) string? s"),
```

Comment:
This doesn't seem right -- the enumerable is an enumerable of `string` and not `string?`, and range variables can't be reassigned so it doesn't make sense if we're trying to allow a later null assignment. No matter what, this test is the "NullableRangeVariableThatIsNonNull" test so the test is no longer matching the original goal, so either the test needs to be renamed or we need to redo the test scenario.

Reply:
This is interesting. It does appear that nullable inference for `var` is not necessary for cases where reassignment is not possible. Several code constructs have this case.

Reply:
@gafter @jasonmalinowski @sharwell I disagree with this. I don't think that whether a variable is read-only or not should affect the inference mechanism, from a language perspective.

In practice, it doesn't matter much. Users will get the same diagnostics either way. They will get a nullable inferred type in QuickInfo with a "variable is not null here", which is correct.

Reply:
Sure, but it still says `string?` which doesn't make any sense. One part of Quick Info might be right, but that doesn't mean it's OK for other parts to make no sense. :smile:

Reply:
> but it still says string? which doesn't make any sense.

I think `string?` is correct, so I'm not sure why it doesn't make any sense :-P

Just to clarify, you'r okay with the change in general, so it makes sense for `var` to infer `string?` in `var x = "";` (ie. we're not dealing with a `foreach`)? 

To me the irregularity of treating `foreach` differently would seem more surprising than the new behavior of `var` in most places. If we accept the new `var` inference, then I don't see why not adapt the behavior generally. :)

Reply:
I'll run this quickly by Mads today if I see him.

Reply:
> I think string? is correct, so I'm not sure why it doesn't make any sense :-P

My concern is a user might see the ? and think that the variable either can be null, or can be assigned null, neither of which is true. The behavior looks like an IDE bug in this case. Minimally, I'd expect that we update this test with a comment explaining this behavior and a link to the LDM tracking issue that was filed so somebody else who has to investigate this is able to get some traction on it.

> Just to clarify, you'r okay with the change in general

I certainly can't say I'm a fan of the LDM direction here; I also certainly know I don't appreciate the tradeoffs being made -- the notes are pretty light here and don't explain the alternatives considered and why they were thrown out.

Reply:
(and my last comment is not in any way a slight to the LDM notes -- capturing a LDM conversation in text form is indeed really difficult. If nothing else, it doesn't mention why `var?` was rejected...)

Reply:
I was strongly against `var?` as it seemed to violate the intent of users who follow the "use var where possible" approach. The style would change to "use var? where possible", which is a longer name _and_ would produce a large amount of churn. The modified definition of `var` itself seems to best reflect the desired behavior for users who prefer types be chosen according to flow analysis, and confusion about the lack of `var` distinguishing between non-null and null value is most likely to come from users who want to _see_ specific types in code, and those users are less likely to be using `var` from the start.

Reply:
I guess this feels far too much like a surprise to be worth it. This makes me distrust var more than be excited to keep using var.